lib/auth/keystore/manager.go

```diff
@@ -0,0 +1,278 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keystore
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509/pkix"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/modules"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+)
+
+// Manager provides an interface to interact with teleport CA private keys,
+// which may be software keys or held in an HSM or other key manager.
+type Manager struct {
+	backend
+}
+
+// backend is an interface that holds private keys and provides signing
+// operations.
+type backend interface {
+	// DeleteUnusedKeys deletes all keys from the KeyStore if they are:
+	// 1. Labeled by this KeyStore when they were created
+	// 2. Not included in the argument usedKeys
+	DeleteUnusedKeys(usedKeys [][]byte) error
+
+	// generateRSA creates a new RSA private key and returns its identifier and
+	// a crypto.Signer. The returned identifier can be passed to getSigner
+	// later to get the same crypto.Signer.
+	generateRSA() (keyID []byte, signer crypto.Signer, err error)
+
+	// getSigner returns a crypto.Signer for the given key identifier, if it is found.
+	getSigner(keyID []byte) (crypto.Signer, error)
+
+	// deleteKey deletes the given key from the KeyStore.
+	deleteKey(keyID []byte) error
+
+	// canSignWithKey returns true if this KeyStore is able to sign with the
+	// given key.
+	canSignWithKey(raw []byte, keyType types.PrivateKeyType) (bool, error)
+}
+
+// Config holds configuration parameters for the keystore.
+type Config struct {
```

Comment:
Are the configs exclusive?
If so, can we please add a comment saying so?

Reply:
Added a comment